File: Data Structures & Algorithms/house-robber/submission-5.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        dp = [0] * len(nums)

        if len(nums) < 2:
            if len(nums) <= 1:
                return nums[0] if nums else 0
            return max(nums[0], nums[1])

        dp[0], dp[1] = nums[0], max(nums[0], nums[1])

        for i in range(2, len(nums)):
            dp[i] = max(dp[i-1], nums[i] + dp[i-2])
        
        return dp[-1]

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-5.java
class Solution {
    // Dynamic Programming

    public int maxProfit(int[] prices) {
        int minBuy = prices[0];
        int maxP = 0;

        for(int sell : prices){
            int profit = sell - minBuy;
            maxP = Math.max(profit, maxP);

            minBuy = Math.min(sell, minBuy);
        }

        return maxP;
    }
}
